addons/block_code/ui/blocks/utilities/template_editor/template_editor.gd

```diff
@@ -0,0 +1,137 @@
+var _parameter_inputs_by_name: Dictionary
+
+@onready var _container := %Container
+@onready var _regex := RegEx.create_from_string(FORMAT_STRING_PATTERN)
+
+
+func _ready() -> void:
+	parent_block = BlockTreeUtil.get_parent_block(self)
+
+	_update_from_format_string()
+
+
+## Set the values of all input parameters based from a dictionary of raw values.
+## Parameters not included in [param raw_values] will be reset to their
+## defaults according to [member parameter_defaults].
+func set_parameter_values(raw_values: Dictionary):
+	for parameter_name in _parameter_inputs_by_name:
+		var parameter_input: ParameterInput = _parameter_inputs_by_name[parameter_name]
+		var parameter_value: Variant = raw_values.get(parameter_name)
+
+		parameter_input.set_raw_input(parameter_value)
+
+
+func get_parameter_values() -> Dictionary:
+	var result: Dictionary
+
+	for parameter_name in _parameter_inputs_by_name:
+		var parameter_input: ParameterInput = _parameter_inputs_by_name[parameter_name]
+		result[parameter_name] = parameter_input.get_raw_input()
+
+	return result
+
+
+func format_statement(statement: String) -> String:
+	return _parameter_inputs_by_name.keys().reduce(_replace_parameter_in_statement, statement)
+
+
+func _replace_parameter_in_statement(statement: String, parameter_name: String) -> String:
+	var parameter_input: ParameterInput = _parameter_inputs_by_name[parameter_name]
+	return statement.replace("{%s}" % parameter_name, parameter_input.get_string())
+
+
+func _update_from_format_string():
+	if not _container:
+		return
+
+	_parameter_inputs_by_name = {}
+	for child in _container.get_children():
+		_container.remove_child(child)
+		child.queue_free()
+
+	var match_id = 0
+	for item in BlockDefinition.parse_display_template(format_string):
+		if item.has("label"):
+			_append_label(item.get("label"))
+		elif item.has("in_parameter"):
+			_append_input_parameter(item.get("in_parameter"), match_id)
+		elif item.has("out_parameter"):
+			_append_output_parameter(item.get("out_parameter"), match_id)
+		match_id += 1
+
+
+func _append_label(label_format: String):
+	var label = Label.new()
+	label.add_theme_color_override("font_color", Color.WHITE)
+	label.text = label_format.strip_edges()
+	_container.add_child(label)
+
+
+func _append_input_parameter(parameter: Dictionary, id: int):
+	var default_value = parameter_defaults.get(parameter["name"])
+
+	var parameter_input: ParameterInput = ParameterInputScene.instantiate()
+	parameter_input.name = "ParameterInput%d" % id
+	parameter_input.placeholder = parameter["name"]
+	parameter_input.variant_type = parameter["type"]
+
+	if default_value is OptionData:
+		var option_data := default_value as OptionData
+		parameter_input.option_data = option_data
+		if option_data.selected < option_data.items.size():
+			parameter_input.default_value = option_data.items[option_data.selected]
+	else:
+		parameter_input.default_value = default_value
+
+	parameter_input.modified.connect(modified.emit)
```

Comment:
It looks like this is causing an error on 4.2 import:

```
SCRIPT ERROR: Parse Error: Cannot find property "emit" on base "Signal".
          at: GDScript::reload (res://addons/block_code/ui/blocks/utilities/template_editor/template_editor.gd:124)
SCRIPT ERROR: Parse Error: Cannot find member "emit" in base "Signal".
          at: GDScript::reload (res://addons/block_code/ui/blocks/utilities/template_editor/template_editor.gd:124)
```
It looks like the way we've been handling this elsewhere is to make a lambda:

```suggestion
	parameter_input.modified.connect(func(): modified.emit())
```